### DISCUSSION TITLE
How to handle corruptRevisionData error?

### DISCUSSION THREAD

**Discussion by kaptnkoala** [cofu-app/cbl-dart#396] [Q&A] [answered]
Sometimes when we are saving a document, that we are updating very frequently, we are receiving a corruptRevisionData error. How should treat this error? From what we can see from the logs, the problem seems to persist, because the same users are running in this error again and again. I have not found any information in the Couchbase Lite docs, what we should we do with a corrupted document like this.

Is it an option to try to delete or purge the document? Or might it be a problem that we are doing changes to a document so frequently?Maybe you have any thoughts or insights on this.

**Comment by blaugold** [comment]
You should be able to update documents as frequently as you want.
If frequent updates lead to data corruption, that is a bug.

@rajagp Is this something you have seen before?
What information does the Couchbase Lite team need to debug this? 
Since this is reported as an error through the C API, there is no stack trace.
I'm curious what could cause a `kCBLErrorCorruptData` error when saving a document?


@kaptnkoala Can you tell that once a document has been corrupted, it is always this document that causes the error?

Deleting the problem document might not work. In Couchbase Lite, deleting is essentially an update to set a flag so that the deletion can be pushed to the server later.
If deleting works, it would also cause the document to be deleted on the server during the next push replication.

Purging only deletes the document locally. If the document exists on the server, it will be pulled again during the next pull replication.
That seems preferable since the user might not lose all the data in the document.

There is an edge case where the purged document is recreated before the revision from the server is pulled.
Recreation meaning a document with the same ID as a purged document is created.
I'm not quite sure how conflict resolution during pull replication would work in this case because the local and remote document share no history. I'll try to find out what the concrete behavior is.

**Comment by kaptnkoala** [comment]
@blaugold From what we can see from the events in our crash reporting tools, it seems that the document remains unaccessible (reading and writing to the document raises the `corruptRevisionData` error), once a document has been corrupted. 

It is our understanding as well, that just deleting the document should not work, because it will just try to create a tombstone like you mentioned in you answer, which is probably not possible in a document state with corrupt revision data. Because the error currently only exists on documents, that are stored locally and not synced with a remote Couchbase instance, there should not be any problem with sync scenarios. 

We have started a release before the weekend, where we try to purge the document, after we received a corruptRevisionData error when writing an affected document. From this change we are now receiving a low amount of crashes, probably because we try to resave the document that we just used for purging. I will attach the Dart code and stack trace, that we see in the crash reporting tools. 

```dart
    try {
      await database.saveDocument(document);
    } catch (error, trace) {
      _logger.error(error, trace, message: 'Failed to update document ${document.id}');
      if (error is DatabaseException && error.code == DatabaseErrorCode.corruptRevisionData) {
        await database.purgeDocument(document);
        await database.saveDocument(document);
      }
    }
```

```
Crashed: DartWorker
0  CouchbaseLite                  0xea27c FL_WipeMemory + 475044
1  CouchbaseLite                  0x1010a8 FL_WipeMemory + 568784
2  CouchbaseLite                  0x709dc FLSlot_SetBlob + 3272
3  CouchbaseLite                  0x71808 CBLDocument_Properties + 12
...
```

**Comment by blaugold** [comment, marked as answer]
When I try to save a purged document like below, I get `DatabaseException(not found, code: notFound)`:

```dart
final doc = MutableDocument();
await db.saveDocument(doc);
await db.purgeDocument(doc);
await db.saveDocument(doc);
```

This comes from CouchbaseLite C so saving a purged document instance is not supported, apparently.
You'll have to recreate the document and continue with the new instance:

```dart
await db.purgeDocument(doc);
doc = MutableDocument.withId(doc.id, doc.toPlainMap());
await db.saveDocument(doc);
return doc;
```

As you mentioned, it probably is not safe to use the corrupted document anyway.

Unfortunately, the symbols in the provided stack trace for the stack frames in `CouchbaseLite` seem to be wrong, except for the first one. The root cause is likely the same one that caused the first stack trace in [this comment](https://github.com/cbl-dart/cbl-dart/issues/390#issuecomment-1185506286). Internally, documents store one or more revisions. The crash occurs because the array of revision is empty in the corrupted document.

**Reply by blaugold** [reply]
PS: `purgeDocument` should work even with the corrupted document because only the id is used.

**Reply by kaptnkoala** [reply]
We have adapted the code like you suggested, release is rolling out today!
Thanks again for your great support!!
